[pythonscripting] Initial contribution

HolgerHees opened the pull request:
Initial release with detailed description in marketplace

https://community.openhab.org/t/python3-scripting/161945

florian-h05 commented:
> and it has still one major known issues https://github.com/openhab/openhab-addons/issues/18054#issuecomment-2634261414

Hasn't his been resolved with the OSGi-ified GraalVM?

HolgerHees commented:
> > and it has still one major known issues [#18054 (comment)](https://github.com/openhab/openhab-addons/issues/18054#issuecomment-2634261414)
> 
> Hasn't his been resolved with the OSGi-ified GraalVM?

yes, you are right. But not for openhab 4... I will remove it here in the description.

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
I am a bit surprised that GraalPython does not provide its own ScriptEngine implementation, but instead you need to create your own around the Polyglot Context.

HolgerHees replied:
yes, that's the case. The ScriptEngine is not really deprecated but not the recommended way of oracle. Oracle recommends to use Polyglot Context directly. 

btw. the GrallPython implementation is inspired a lot by the official JS Script Engine implementation, but in a cleaned and simplified variant.

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,508 @@
+        // ==> ThreadGroup is closed and new python threads can't started anymore
+        // - in jdk21 is handled differently
+        // ==> ThreadGroups does not use the daemon flag anymore and ThreadGroups are not closable anymore
+        // - can be removed in jdk21 only environments
```
Can we remove this here hence?
main branch is openHAB 5 which requires JDK 21. I know you want to provide a 4.x version of the add-on, but this ideally shouldn't pollute the code on main.

HolgerHees replied:
currently, if I build the openhab4 version, I just copy over the source code and compile. If this is a blocker for you, I can remove it.

My current plant is to keep it until openhab 5 is released. Until this time I will also compile specific openhab4 versions. So I would remove it in a openhab 5.1 branch.

florian-h05 replied:
I would prefer to remove it here, and then add it when compiling the 4.x version. You can keep a copy of the 4.3.x branch with the Python add-on and an extra commit containing this change, and whenever the Python binding changes on main, rebase it.

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,508 @@
+                        wrapperContent, PythonScriptEngineFactory.PYTHON_WRAPPER_FILE_PATH.toString()).build());
+
+                // inject scope, Registry and logger
+                if (!pythonScriptEngineConfiguration.isInjection(PythonScriptEngineConfiguration.INJECTION_DISABLED)
```
I really like that one can control where to automatically inject the library, that makes much sense. I will probably add this to JS Scripting as well.

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,508 @@
+
+    @Override
+    protected Exception afterThrowsInvocation(Exception e) {
+        // OPS4J Pax Logging holds a reference to the exception, which causes the OpenhabGraalJSScriptEngine to not be
```
Probably update my comment:

```suggestion
        // OPS4J Pax Logging holds a reference to the exception, which causes the PythonScriptEngine to not be
```

HolgerHees replied:
fixed

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+This add-on provides support for Python 3 that can be used as a scripting language within automation rules.
+It is based on [GraalPy](https://www.graalvm.org/python/) from the [GraalVM project](https://www.graalvm.org/).
+
+Also included is [openhab-python](https://github.com/HolgerHees/openhab-python), a python library to support automation in openHAB. It provides convenient access to common core openHAB functions that make the full range of Java APIs easily accessible and usable
```
```suggestion
Also included is [openhab-python](https://github.com/HolgerHees/openhab-python), a python library to support automation in openHAB. It provides convenient access to common core openHAB functions that make the full range of Java APIs easily accessible and usable.
```

HolgerHees replied:
fixed

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/README.md:
````diff
@@ -0,0 +1,495 @@
+Alternatively, you can create scripts in the `automation/python` configuration directory. If you create an empty file called `test.py`, you will see a log line with information similar to:
+
+```text
+... [INFO ] [ort.loader.AbstractScriptFileWatcher]] - (Re-)Loading script '/openhab/conf/automation/python/test.py'
````
```suggestion
... [INFO ] [ort.loader.AbstractScriptFileWatcher] - (Re-)Loading script '/openhab/conf/automation/python/test.py'
```

HolgerHees replied:
fixed

florian-h05 commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+:::
+ 
+
+## `PY3` Transformation
```
Once this is merged, we can update the UI provide doc links, code snippets etc, see https://github.com/openhab/openhab-webui/blob/main/bundles/org.openhab.ui/web/src/assets/transformations.js.

florian-h05 replied:
@HolgerHees Do you want to do this yourself?

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/bnd.bnd:
```diff
@@ -0,0 +1,22 @@
+ org.graalvm.sdk.nativeimage;bundle-version="24.2.0",\
+ org.graalvm.sdk.word;bundle-version="24.2.0",\
+ org.graalvm.shadowed.icu4j;bundle-version="24.2.0",\
+ org.graalvm.shadowed.xz;bundle-version="24.2.0",\
```
is cleaned

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/pom.xml:
```diff
@@ -0,0 +1,263 @@
+            <configuration>
+              <connectionUrl>scm:git:https://github.com/openhab/openhab-python</connectionUrl>
+              <checkoutDirectory>${project.build.directory}/python</checkoutDirectory>
+              <scmVersion>main</scmVersion>
```
is now fixed to v1.0.0-rc1

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngineFactory.java:
```diff
@@ -0,0 +1,206 @@
+/**
+ * This is an implementation of {@link ScriptEngineFactory} for Python.
+ *
+ * @author Holger Hees - initial contribution
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonBindings.java:
```diff
@@ -0,0 +1,126 @@
+import org.graalvm.polyglot.Value;
+
+/***
+ * @author Holger Hees - Initial contribution
```
solved

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,104 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```
is fixed...

The javascript version was my inspiration for this. But in the latest version, only the class name is similar :-)

I think, still a lot of code is somehow similar to the javascript implementation. Maybe in the future we could separate thing to share in a common scripting package. I mean not this class, but there are still other candidates.

florian-h05 replied:
Yeah, the only "real" place for shared code is core.

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,13 @@
+
+	<type>automation</type>
+	<name>Python Scripting</name>
+	<description>This adds a Python script engine.</description>
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,76 @@
+			<label>Use scope and import wrapper</label>
+			<description><![CDATA[
+			This enables a scope module and and import wrapper.<br>
+			An scope module is an encapsulated module containing all openhab jsr223 objects and can be imported with 'import scope'<br>
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,508 @@
+    public static final String LOGGER_INIT_NAME = "__logger_init__";
+
+    /** Shared Polyglot {@link Engine} across all instances of {@link PythonScriptEngine} */
+    private static final Engine ENGINE = Engine.newBuilder().allowExperimentalOptions(true)
```
This is just a static generic polyglot engine shared across all GraalPythonScriptEngine instances.

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,508 @@
+            .option("engine.WarnInterpreterOnly", "false").build();
+
+    /** Provides unlimited host access as well as custom translations from Python to Java Objects */
+    private static final HostAccess HOST_ACCESS = HostAccess.newBuilder(HostAccess.ALL)
```
yes, I learned a lot from javascript binding :-)

florian-h05 reviewed (commented):
LGTM for me now, only two things left:

- https://github.com/openhab/openhab-addons/pull/18499#discussion_r2035234856-
- Please generate the default translations: `mvn i18n:generate-default-translations -pl :org.openhab.automation.pythonscripting`

florian-h05 reviewed (approved):
LGTM now, thanks you very much!

@openhab/add-ons-maintainers Can you please review?

wborn reviewed (changes requested):
Many thanks for the PR! 👍 I added a few comments below:

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/pom.xml:
```diff
@@ -0,0 +1,263 @@
+  <name>openHAB Add-ons :: Bundles :: Automation :: Python Scripting</name>
+
+  <properties>
+    <bnd.importpackage>!sun.misc.*,
+      !sun.reflect.*,
+      !com.sun.management.*,
+      !jdk.internal.reflect.*,
+      !jdk.vm.ci.services</bnd.importpackage>
```
I'm not sure about this. I just used the jsscripting pom file as a template.

ccutrer replied:
No, it is not necessary. I just tested with my Jinja PR, and removing this property and the JAR still works in openHAB.

HolgerHees replied:
for me, this does not work. I tested it right now in openhab4.

If I compile a jar **with** this code snipped everything works.

If I compile a jar **without** this code snipped, I get the following error

```
Error while starting bundle: file:/openhab/addons/org.openhab.automation.pythonscripting-4.3.3-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.automation.pythonscripting [351]
  Unresolved requirement: Import-Package: com.oracle.truffle.api
```

I verified it several times and the only difference between both jar files is the missing "bnd.importpackage" setting.

florian-h05 replied:
You can set the Import-Package directive in the bnd file and remote it from the POMY

HolgerHees replied:
this works! it is fixed/cleaned now!

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+# Python 3 Binding
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+
+### decorator @rule
+
+the decorator will register the decorated class as a rule. It will wrap and extend the class with the following functionalities
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,428 @@
+        final StringBuilder builder = new StringBuilder();
+        final char[] buffer = new char[1024];
+        try {
+            try {
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,428 @@
+    private Context createContext(Bindings engineB) {
+        Object ctx = engineB.get(POLYGLOT_CONTEXT);
+        if (!(ctx instanceof Context)) {
+            Context.Builder builder = contextConfig;
+            ctx = createDefaultContext(builder, context);
```
is fixed. was a leftover.

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngineFactory.java:
```diff
@@ -0,0 +1,206 @@
+                    .findEntries(resourceLibPath, "*.py", true);
+
+            if (Files.exists(PythonScriptEngineFactory.PYTHON_OPENHAB_LIB_PATH)
+                    && Files.list(PYTHON_OPENHAB_LIB_PATH).count() > 0) {
```
is fixed

HolgerHees commented:
> Many thanks for the PR! 👍 I added a few comments below:

Thanks for your feedback.  I fixed almost everything related to your review, except 3 things where I have open questions how to proceed.

ccutrer commented on bundles/org.openhab.automation.pythonscripting/pom.xml:
```diff
@@ -0,0 +1,263 @@
+                <excludes>
+                  <exclude>org.lastnpe.eea:eea-all</exclude>
+                  <exclude>org.apache.karaf.features:framework</exclude>
+                  <!-- we use OSGI-ified version, so we don't need the following -->
```
It's me, it's me! And I spent considerable time working on it today. I've got a PR for doing it in just the Jinja transformation (because building a Home Assistant jar is more complicated, because it's a binding comprised of multiple bundles). I used your pom.xml here as an example, though I did flip this around to be in includes, instead of excludes. That eliminated several more JARs from the shaded JAR. And I've got it working... with one massive exception - unit tests don't work! See the build associated with https://github.com/openhab/openhab-addons/pull/18590. I get that same issue locally. AFAICT, it's something to do with when GraalPy loads bouncycastle, and somehow that JAR is signed, but it can't verify the signature. I've tried various methods of disabling signature checking, but it hasn't worked. Do either of you have any suggestions for me?

I also spent some time trying to OSGi-ify more of the Graal dependencies so that there won't be 100+ MB of duplicated jars between these three addons, but I was unsuccessful. I was able to get the the Jinja add-on down to like 11MB or something with all the Graal dependencies OSGi-ified, and the bundles successfully loaded inside openHAB, but when the Jinja add-on tries to instantiate the polyglot Context, it errors that Polyglot couldn't find any implementations:

```
Caused by: java.lang.IllegalStateException: No language and polyglot implementation was found on the module-path. Make sure at last one language is added to the module-path. 
        at org.graalvm.polyglot.Engine$PolyglotInvalid.noPolyglotImplementationFound(Engine.java:1921) ~[?:?]
        at org.graalvm.polyglot.Engine$PolyglotInvalid.newDefaultFileSystem(Engine.java:1945) ~[?:?]
        at org.graalvm.polyglot.io.FileSystem.newDefaultFileSystem(FileSystem.java:501) ~[?:?]
        at org.graalvm.python.embedding.VirtualFileSystemImpl.<init>(VirtualFileSystemImpl.java:332) ~[?:?]
        at org.graalvm.python.embedding.VirtualFileSystem.<init>(VirtualFileSystem.java:273) ~[?:?]
        at org.graalvm.python.embedding.VirtualFileSystem$Builder.build(VirtualFileSystem.java:254) ~[?:?]
        at org.openhab.transform.jinja.internal.JinjaTransformationService.<init>(JinjaTransformationService.java:51) ~[?:?]
```

I tried poking around for a while on this too. I'm pretty sure [this](https://github.com/oracle/graal/blob/master/sdk/src/org.graalvm.polyglot/src/org/graalvm/polyglot/Engine.java#L1853) is where Polyglot tries to load things, but I couldn't figure out a way to get `ServiceLoader` to work since Karaf gives every bundle its own ClassLoader and Module. There doesn't seem to be any way to dynamically register a service onto a class loader or module, or tell any two existing modules to "be friends". So maybe the the secret is we need to manually add the proper service file to the polyglot osgi-ified bundle, and an OSGi level import on the bundle that provides said service? I'm not sure how well that would work to provide JavaScript OR Python from Graal, without requiring both for all addons that use either. Anyhow, I'm done for the day, and welcome any suggestions on where to go next for either issue.

HolgerHees replied:
I would recommend to keep it like it is for now, to avoid release blocking problems. 

For the next openhab version like 5.x we could continue to work on this. And I think we need an OSGI expert to help us here.

florian-h05 replied:
> So maybe the the secret is we need to manually add the proper service file to the polyglot osgi-ified bundle, and an OSGi level import on the bundle that provides said service?

I was also already think about that, as we definitely need to tell the polyglot bundle which language implementations are available. We could definitely try it, if the feature.xml of the add-on requires the implementation the Implantation should be available via OSGi when the polyglot bundle tries to use it. The bigger question is what does the polyglot bundle do if some of the implantations listed are not available as the bundle is not installed and not used, will it anyway try to load them or does it load them only on demand.

Another option would be to create different OSGi-ified „flavours“ of the polyglot bundle for the different language and with the matching service loading configs. Since the polyglot bundle itself is relatively small it would be cheap to have multiple flavours of it.

ccutrer replied:
> Another option would be to create different OSGi-ified „flavours“ of the polyglot bundle for the different language and with the matching service loading configs. Since the polyglot bundle itself is relatively small it would be cheap to have multiple flavours of it.

Ooh, I like that idea better than the other idea I thought of last night - to have a "fat" OSGi-ified bundle of all the necessary JARs, but it would mean either having both a JS bundle and a Python bundle with overlap, or a single JS+Python bundle with (a lot of!) possibly unused extras.

wborn replied:
Maybe someone can create an issue for reusing those bundles? Then I'll resolve this comment and hopefully the respective add-ons shrink in size after that issue is resolved. :slightly_smiling_face:

HolgerHees replied:
There is already a pull request (DRAFT)

https://github.com/openhab/openhab-osgiify/pull/58

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+
+### module scope
+
+The scope module encapsulates all [default jsr223 objects/presents](https://www.openhab.org/docs/configuration/jsr223.html#default-preset-importpreset-not-required) into a new object. You can use it like below
```
same arguments as in my first comment related to {{base}} usage

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
````diff
@@ -0,0 +1,495 @@
+print(scope.ON)
+```
+
+You can also import additional [jsr223 presents](https://www.openhab.org/docs/configuration/jsr223.html#rulesimple-preset) like
````
same arguments as in my first comment related to {{base}} usage

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
````diff
@@ -0,0 +1,495 @@
+from scope import osgi
+```
+
+Additionally you can import all java classes from 'org.openhab' package like
````
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+
+## Others
+
+### python <=> java conversion
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,500 @@
+    /**
+     * Creates an implementation of ScriptEngine {@code (& Invocable)}, wrapping the contained engine,
+     * that tracks the script lifecycle and provides hooks for scripts to do so too.
+     *
+     * @param pythonDependencyTracker
+     * @param injectionEnabled
+     * @param scopeEnabled
+     * @param cachingEnabled
+     * @param jythonEmulation
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,500 @@
+    // collect JSR223 (scope) variables separately, because they are delivered via 'import scope'
+    public void put(String key, Object value) {
+        if ("javax.script.filename".equals(key)) {
+            // super.put("__file__", value);
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,500 @@
+    @Override
+    public boolean tryLock(long l, TimeUnit timeUnit) throws InterruptedException {
+        boolean acquired = lock.tryLock(l, timeUnit);
+        if (acquired) {
+            logger.debug("Lock acquired.");
+        } else {
+            logger.debug("Lock not acquired.");
+        }
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,428 @@
+
+    GraalPythonScriptEngine(GraalPythonScriptEngineFactory factory, Engine engine, Context.Builder contextConfig) {
+        Engine engineToUse = (engine != null) ? engineToUse = engine : factory.getPolyglotEngine();
+        // this.factory = (factory == null) ? new GraalPythonScriptEngineFactory(engineToUse) : factory;
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,101 @@
+
+    private final Logger logger = LoggerFactory.getLogger(ScriptExtensionModuleProvider.class);
+
+    private Map<String, Object> globals = new HashMap<String, Object>();
```
fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,495 @@
+
+openHAB provides several [data transformation services](https://www.openhab.org/addons/#transform) as well as the script transformations, that are available from the framework and need no additional installation.
+It allows transforming values using any of the available scripting languages, which means Python Scripting is supported as well.
+See the [transformation docs](https://openhab.org/docs/configuration/transformations.html#script-transformation) for more general information on the usage of script transformations.
```
If I was to maintain it I would probably automate generating this README using some script (or Maven) so it will be easier to update the header and replace the URLs.

Note that line 20 is already using `{{base}}` which triggered me to comment on this.

HolgerHees replied:
I remove the usage of {{base}} for now. But I understand you objections. I'll definitely work on a solution after the 5.0 release. For now, I'd like to keep things simple.

wborn replied:
I created https://github.com/openhab/openhab-docs/pull/2488 so it will be handled in the documentation build. :slightly_smiling_face:

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
````diff
@@ -0,0 +1,574 @@
+   calc(input)
+   ```
+
+2. Using `PY3(<scriptname>.py):%s` as Item state transformation.
+3. Passing parameters is also possible by using a URL like syntax: `PY3(<scriptname>.py?arg=value)`.
+   Parameters are injected into the script and can be referenced like variables.
+
+Simple transformations can aso be given as an inline script: `PY3(|...)`, e.g. `PY3(|"String has " + input.length + "characters")`.
````
```suggestion
2. Using `PY3(<scriptname>.py):%s` as Item state transformation.
3. Passing parameters is also possible by using a URL like syntax: `PY3(<scriptname>.py?arg=value)`.
   Parameters are injected into the script and can be referenced like variables.

Simple transformations can also be given as an inline script: `PY3(|...)`, e.g. `PY3(|"String has " + input.length + "characters")`.
```

HolgerHees replied:
So what is your recommendation here? I like to have PY for the new one and PY2 for the old one.

My recommendation is to keep it like it is for now.

And if this binding is merged. I will create a new pull request where I change the jython one to PY2 and PY for this addon.

wborn replied:
Yes it's better to do this in a follow up PR so CI only builds this add-on when commits are added instead of all add-ons.

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,490 @@
+    @Override
+    public boolean tryLock() {
+        boolean acquired = lock.tryLock();
+        if (acquired) {
+            logger.debug("Lock acquired.");
+        } else {
+            logger.debug("Lock not acquired.");
+        }
```
is fixed. Was just a mistake that I forgot this place.

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/scope/AbstractScriptExtensionProvider.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Override
+    public Collection<String> getDefaultPresets() {
+        return Collections.emptyList();
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/scope/AbstractScriptExtensionProvider.java:
```diff
@@ -0,0 +1,96 @@
+            return results;
+        }
+
+        return Collections.emptyMap();
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,424 @@
+    }
+
+    GraalPythonScriptEngine(GraalPythonScriptEngineFactory factory, Engine engine, Context.Builder contextConfig) {
+        Engine engineToUse = (engine != null) ? engineToUse = engine : factory.getPolyglotEngine();
```
sometimes you are blind to your own nonsense :-)

is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,101 @@
+
+    private Map<String, Object> globals = new HashMap<>();
+
+    public ScriptExtensionModuleProvider() {
+    }
+
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+# Python 3 Scripting
```
is fixed.

I was thinking first to permanently name it python 3, to be prepared for the future :-)

HolgerHees commented:
@wborn Thanks again! I fixed most of your comments... the rest is answered.

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+
+| Function                 | Usage                                                                                 | Description                                                                                         |
+| ------------------------ | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- |
+| getStableMinMaxState     | getStableMinMaxState(time_slot, end_time = None)                                      | Average calculation which takes into account the values ​​depending on their duration                 |
+| getStableState           | getStableState(time_slot, end_time = None)                                            | Average calculation which takes into account the values ​​depending on their duration                 |
```
Why the zero width spaces?

```suggestion
| getStableMinMaxState     | getStableMinMaxState(time_slot, end_time = None)                                      | Average calculation which takes into account the values depending on their duration                 |
| getStableState           | getStableState(time_slot, end_time = None)                                            | Average calculation which takes into account the values depending on their duration                 |
```

HolgerHees replied:
![grafik](https://github.com/user-attachments/assets/f09bb80b-55bf-47d2-924b-afb959f7f796)

it is just to align table column separator which makes it easier to read during edit.

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
````diff
@@ -0,0 +1,574 @@
+Timer.createTimeout(60, test)
+```
+
+Below is a complex example of 2 sensor values ​​that are expected to be transmitted in a certain time window (e.g. one after the other).
````
More zero width spaces?

```suggestion
Below is a complex example of 2 sensor values that are expected to be transmitted in a certain time window (e.g. one after the other).
```

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+
+There are 3 ways of logging.
+
+1. using normal print statements. In this case they are redirected to the default openhab logfile and marked with log level INFO or ERROR
```
```suggestion
1. using normal print statements. In this case they are redirected to the default openHAB logfile and marked with log level INFO or ERROR
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+| sendCommandIfDifferent   | sendCommandIfDifferent(command)                                                       |                                                                                                     |
+| getPersistence           | getPersistence(service_id = None)                                                     | [ItemPersistence](#class-itempersistence)                                         |
+| getSemantic              | getSemantic()                                                                         | [ItemSemantic](#class-itemsemantic)                                               |
+| <...>                    | see [openHAB Item api](https://www.openhab.org/javadoc/latest/org/openhab/core/items/item) |                                                                                                |
```
```suggestion
| <...>                    | see [openHAB Item API](https://www.openhab.org/javadoc/latest/org/openhab/core/items/item) |                                                                                                |
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,424 @@
+    private static final String PYTHON_OPTION_FORCEIMPORTSITE = "python.ForceImportSite";
+    private static final String PYTHON_OPTION_CHECKHASHPYCSMODE = "python.CheckHashPycsMode";
+
+    static final String MAGIC_OPTION_PREFIX = "polyglot.py.";
```
This doesn't seem to be used?

```suggestion
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/scriptengine/helper/LifecycleTracker.java:
```diff
@@ -0,0 +1,41 @@
+ */
+@NonNullByDefault
+public class LifecycleTracker {
+    List<Runnable> disposables = new ArrayList<>();
```
```suggestion
    private List<Runnable> disposables = new ArrayList<>();
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/scriptengine/helper/LifecycleTracker.java:
```diff
@@ -0,0 +1,41 @@
+ */
+@NonNullByDefault
+public class LifecycleTracker {
+    List<Runnable> disposables = new ArrayList<>();
```
```suggestion
    private List<Runnable> disposables = new ArrayList<>();
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,98 @@
+
+    public static final String IMPORT_PROXY_NAME = "__import_proxy__";
+
+    public static final String OENHAB_MODULE_PREFIX = "org.openhab";
```
```suggestion
    public static final String OPENHAB_MODULE_PREFIX = "org.openhab";
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,98 @@
+    public ModuleLocator locatorFor(Context ctx, String engineIdentifier,
+            ScriptExtensionAccessor scriptExtensionAccessor) {
+        return (name, fromlist) -> {
+            Map<String, Object> symbols = new HashMap<String, Object>();
```
```suggestion
            Map<String, Object> symbols = new HashMap<>();
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,98 @@
+        return (name, fromlist) -> {
+            Map<String, Object> symbols = new HashMap<String, Object>();
+            if (name.startsWith(OENHAB_MODULE_PREFIX)) {
+                List<String> classList = new ArrayList<String>();
```
```suggestion
                List<String> classList = new ArrayList<>();
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,98 @@
+            } else if (name.startsWith(SCOPE_MODULE_PREFIX)) {
+                String[] segments = name.split("\\.");
+                if (name.equals(SCOPE_MODULE_PREFIX)) {
+                    Map<String, Object> possibleSymbols = new HashMap<String, Object>(this.globals);
```
```suggestion
                    Map<String, Object> possibleSymbols = new HashMap<>(this.globals);
```

HolgerHees replied:
is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/src/main/resources/OH-INF/i18n/pythonscripting.properties:
```diff
@@ -0,0 +1,26 @@
+automation.config.pythonscripting.jythonEmulation.label = Enable Jython emulation
+automation.config.pythonscripting.jythonEmulation.description = This enables Jython emulation in GraalPy. It is strongly recommended to update code to GraalPy and Python 3 as the emulation can have performance degradation. For tips and instructions, please refer to <a href="https://www.graalvm.org/latest/reference-manual/python/Modern-Python-on-JVM">Jython Migration Guide</a>.
+automation.config.pythonscripting.scopeEnabled.label = Use scope and import wrapper
+automation.config.pythonscripting.scopeEnabled.description = This enables a scope module and and import wrapper.<br> An scope module is an encapsulated module containing all openhab jsr223 objects and can be imported with <code>import scope</code><br> Additionally you can run an import like <code>from org.openhab.core import OpenHAB</code>
```
```suggestion
automation.config.pythonscripting.scopeEnabled.description = This enables a scope module and and import wrapper.<br> An scope module is an encapsulated module containing all openHAB jsr223 objects and can be imported with <code>import scope</code><br> Additionally you can run an import like <code>from org.openhab.core import OpenHAB</code>
```

HolgerHees replied:
is a generated file. But the root cause in OH-INF/config/config.xml is fixed

wborn commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+
+### limitations
+
+- graalby can't handle arguments in constructors of java objects. Means you can't instantiate a javaobject in python with a parameter. https://github.com/oracle/graalpython/issues/367
+- graalpy does not really support python 'set' types as arguments of function calls to java objects https://github.com/oracle/graalpython/issues/260
```
```suggestion
- GraalPy can't handle arguments in constructors of java objects. Means you can't instantiate a Java object in Python with a parameter. https://github.com/oracle/graalpython/issues/367
- GraalPy does not really support python 'set' types as arguments of function calls to Java objects https://github.com/oracle/graalpython/issues/260
```

HolgerHees replied:
is fixed

wborn reviewed (changes requested):
After scrolling through the code again I added a few more (minor) comments:

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+
+| Class                    | Usage                                                                                 | Description                                                                                         |
+| ------------------------ | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- |
+| Audio                    | see [openhab Audio api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/audio)      |                                                                          |
+| BusEvent                 | see [openhab BusEvent api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/busevent) |                                                                         |
+| Ephemeris                | see [openhab Ephemeris api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/ephemeris) |                                                                       |
+| Exec                     | see [openhab Exec api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/exec)        | e.g. Exec.executeCommandLine(timedelta(seconds=1), "whoami")             |
+| HTTP                     | see [openhab HTTP api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/http)        |                                                                          |
+| Log                      | see [openhab Log api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/log)          |                                                                          |
+| Ping                     | see [openhab Ping api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/ping)        |                                                                          |
+| ScriptExecution          | see [openhab ScriptExecution api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/scriptexecution) |                                                           |
+| Semantic                 | see [openhab Semantic api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/semantic) |                                                                         |
+| ThingAction              | see [openhab ThingAction api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/things) |                                                                   |
+| Transformation           | see [openhab Transformation api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/transformation) |                                                             |
+| Voice                    | see [openhab Voice api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/voice)      |                                                                          |
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+| ------------------------ | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- |
+| getStableMinMaxState     | getStableMinMaxState(time_slot, end_time = None)                                      | Average calculation which takes into account the values ​​depending on their duration                 |
+| getStableState           | getStableState(time_slot, end_time = None)                                            | Average calculation which takes into account the values ​​depending on their duration                 |
+| <...>                    | see [openHAB PersistenceExtensions api](https://www.openhab.org/javadoc/latest/org/openhab/core/persistence/extensions/persistenceextensions) |                                             |
+
+### class ItemSemantic 
+
+ItemSemantic is a wrapper around [openHAB Semantics](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/semantics). The parameters 'item', as part of the Wrapped Java API, is not needed because it is inserted automatically.
+
+| Function                 | Usage                                                                                 |
+| ------------------------ | ------------------------------------------------------------------------------------- |
+| <...>                    | see [openHAB Semantics api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/semantics) |
+
+### class Thing 
+
+Thing is a wrapper around [openHAB Thing](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/thing). 
+
+| Function                 | Usage                                                                                 |
+| ------------------------ | ------------------------------------------------------------------------------------- |
+| <...>                    | see [openHAB Thing api](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/thing) |
+
+### class Channel 
+
+Channel is a wrapper around [openHAB Channel](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/type/channelgrouptype). 
+
+| Function                 | Usage                                                                                 |
+| ------------------------ | ------------------------------------------------------------------------------------- |
+| <...>                    | see [openHAB Channel api](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/type/channelgrouptype) |
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+
+### class Set
+
+This is a helper class which makes it possible to use a python 'set' as an argument for java class method calls
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/README.md:
```diff
@@ -0,0 +1,574 @@
+
+#### Exception during helper lib initialisation
+
+There were problems during the deployment of the helper libs. A typical error is an insufficient permission. The folder "conf/automation/python/" must be writeable by openhab.
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,486 @@
+    }
+
+    private static Set<String> transformGraalWrapperSet(Value value) {
+        Set<String> set = new HashSet<String>();
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,424 @@
+
+    private final Logger logger = LoggerFactory.getLogger(GraalPythonScriptEngine.class);
+
+    interface MagicBindingsOptionSetter {
+
+        String getOptionKey();
+
+        Context.Builder setOption(Builder builder, Object value);
+    }
+
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngineFactory.java:
```diff
@@ -0,0 +1,167 @@
+        super();
+        this.userDefinedEngine = null;
+
+        defaultEngine = new WeakReference<Engine>(createDefaultEngine());
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/scriptengine/InvocationInterceptingScriptEngineWithInvocableAndCompilableAndAutoCloseable.java:
```diff
@@ -0,0 +1,196 @@
+        }
+    }
+
+    private CompiledScript wrapCompiledScript(CompiledScript script) throws ScriptException {
```
is fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/scriptengine/helper/LifecycleTracker.java:
```diff
@@ -0,0 +1,41 @@
+public class LifecycleTracker {
+    List<Runnable> disposables = new ArrayList<>();
+
+    public void addDisposeHook(Runnable disposable) {
+        disposables.add(disposable);
+    }
+
```
To give the full context...

The original Lifecycle Tracker does not work for graalvm, because it is disposed after the graalvm is closed. This means that no dispose script can be executed.

For that reason I replace the scope injected lifecycle tracker with my own implementation. I duplicated also the API, to be prepared to switch back if the original lifecycle tracker is fixed/refactored sometimes.

https://github.com/HolgerHees/openhab-addons/blob/8bd275ad0b658b123f9f8f272a4cc706db732db9/bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java#L358

The usage of "addDisposeHook" is almost inside the python code

e.g. https://github.com/openhab/openhab-python/blob/50030405229fc5d91ea3859940ec5fb6879c7d7f/src/helper.py#L614

it register some python scripts which should be executed on script shutdown events.

HolgerHees commented:
@wborn and again, a big thank you :-) I'm looking forward to the moment when all of that is merged!

wborn reviewed (approved):
Many thanks for getting it all to work and mergeable @HolgerHees! :+1: 
Let's start enjoying it in the snapshot builds. :rocket:

jlaur commented:
@HolgerHees - congrats! Thanks for the contribution, @HolgerHees. It was a joy see you and @wborn go through the quick iterations to get this to the finish line. 👍
